File: src/report.rs
//! Rendering of the human readable output.

use std::ffi::{OsStr, OsString};
use std::path::Path;

use crate::color::{Painter, Style};

/// The width of the rules that frame the sections.
pub const HEADER_WIDTH: usize = 80;

/// The character set that is used for the decorative glyphs.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GlyphSet {
    /// Emoji and box drawing characters.
    Unicode,
    /// Plain ASCII replacements for terminals that cannot render more.
    Ascii,
}

/// The glyphs of one character set.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Glyphs {
    /// The prefix of the repository line, for example `📂 `.
    pub folder: &'static str,
    /// The separator between the repository and the branch, for example `│`.
    pub branch_sep: &'static str,
    /// The marker of a failing repository.
    pub failure: &'static str,
    /// The character used for the thin rule of the summary.
    pub rule: char,
}

impl Glyphs {
    /// The Unicode glyphs.
    pub const UNICODE: Self = Self {
        folder: "📂 ",
        branch_sep: "│",
        failure: "✗",
        rule: '─',
    };

    /// The ASCII glyphs.
    pub const ASCII: Self = Self {
        folder: "",
        branch_sep: "|",
        failure: "x",
        rule: '-',
    };

    /// The glyphs of a character set.
    pub fn for_set(set: GlyphSet) -> Self {
        match set {
            GlyphSet::Unicode => Self::UNICODE,
            GlyphSet::Ascii => Self::ASCII,
        }
    }
}

/// One repository that failed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Failure {
    /// The directory name of the repository.
    pub name: OsString,
    /// The exit code that git reported.
    pub exit_code: i32,
}

/// The `====` framed header of a repository.
pub fn repo_header(name: &OsStr, branch: &str, paint: &Painter, glyphs: &Glyphs) -> String {
    let rule = "=".repeat(HEADER_WIDTH);
    let mut out = String::with_capacity(4 * HEADER_WIDTH);
    out.push_str(&paint.paint(Style::Cyan, &rule));
    out.push('\n');
    out.push_str(&paint.paint(Style::Cyan, &format!("{}Folder: ", glyphs.folder)));
    out.push_str(&paint.paint_os(Style::BoldYellow, name));
    out.push(' ');
    out.push_str(&paint.paint(Style::Cyan, &format!("{} Branch: ", glyphs.branch_sep)));
    out.push_str(&paint.paint(Style::BoldGreen, branch));
    out.push('\n');
    out.push_str(&paint.paint(Style::Cyan, &rule));
    out.push('\n');
    out
}

/// The counts of one run.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Summary<'a> {
    /// How many repositories finished successfully.
    pub succeeded: usize,
    /// How many repositories were not visited, for example with `--fail-fast`.
    pub skipped: usize,
    /// The repositories that failed.
    pub failures: &'a [Failure],
}

/// The closing summary block.
pub fn summary_block(summary: &Summary<'_>, paint: &Painter, glyphs: &Glyphs) -> String {
    let total = summary.succeeded + summary.skipped + summary.failures.len();
    let rule: String = std::iter::repeat_n(glyphs.rule, HEADER_WIDTH).collect();
    let mut counts = vec![format!(
        "{total} {}, {} succeeded",
        plural(total, "repository", "repositories"),
        summary.succeeded
    )];
    if !summary.failures.is_empty() {
        counts.push(format!("{} failed", summary.failures.len()));
    }
    if summary.skipped > 0 {
        counts.push(format!("{} skipped", summary.skipped));
    }

    let mut out = String::new();
    out.push_str(&paint.paint(Style::Cyan, &rule));
    out.push('\n');

    let style = if summary.failures.is_empty() {
        Style::BoldGreen
    } else {
        Style::Red
    };
    out.push_str(&paint.paint(style, &format!("mgit: {}", counts.join(", "))));
    out.push('\n');

    for failure in summary.failures {
        out.push_str("  ");
        out.push_str(&paint.paint(
            Style::Red,
            &format!("{} {}", glyphs.failure, display_name(&failure.name)),
        ));
        out.push_str(&format!(" (exit code {})\n", failure.exit_code));
    }

    out
}

/// The message that is printed when no repository was found.
pub fn missing_repository_warning(root: &Path, depth: usize) -> String {
    format!(
        "mgit: no Git repository found in `{}` (searched {} {})\n",
        root.display(),
        depth,
        plural(depth, "level", "levels")
    )
}

fn display_name(name: &OsStr) -> String {
    name.to_string_lossy().into_owned()
}

fn plural(count: usize, singular: &str, plural: &str) -> String {
    if count == 1 {
        singular.to_owned()
    } else {
        plural.to_owned()
    }
}

File: website/assets/js/engine/report.js
// Rendering of the human readable output, ported from src/report.rs and
// src/color.rs. The output is produced as text with the same escape sequences
// the real program writes, so the parity tests can compare it character for
// character and the terminal view can parse the very same stream.

export const HEADER_WIDTH = 80;

const ESC = '\x1b';
const RESET = ESC + '[0m';

/** The ANSI styles mgit uses, with the escape sequence of each. */
export const STYLE = {
  plain: null,
  cyan: ESC + '[0;36m',
  boldYellow: ESC + '[1;33m',
  boldGreen: ESC + '[1;32m',
  red: ESC + '[0;31m',
  bold: ESC + '[1m',
};

/** The Unicode glyph set. */
export const UNICODE_GLYPHS = { folder: '\u{1F4C2} ', branchSep: '│', failure: '✗', rule: '─' };
/** The ASCII glyph set, for terminals that cannot render more. */
export const ASCII_GLYPHS = { folder: '', branchSep: '|', failure: 'x', rule: '-' };

/** Writes text with ANSI escape sequences, or verbatim when color is off. */
export class Painter {
  constructor(enabled) {
    this.enabled = !!enabled;
  }

  paint(style, text) {
    const code = STYLE[style];
    if (!this.enabled || !code) return text;
    return code + text + RESET;
  }
}

/** The `====` framed header of one repository. */
export function repoHeader(name, branch, painter, glyphs) {
  const rule = '='.repeat(HEADER_WIDTH);
  return (
    painter.paint('cyan', rule) + '\n' +
    painter.paint('cyan', glyphs.folder + 'Folder: ') +
    painter.paint('boldYellow', name) +
    ' ' +
    painter.paint('cyan', glyphs.branchSep + ' Branch: ') +
    painter.paint('boldGreen', branch) + '\n' +
    painter.paint('cyan', rule) + '\n'
  );
}

/**
 * The closing summary block.
 * @param {{succeeded: number, skipped: number, failures: {name: string, exitCode: number}[]}} summary
 */
export function summaryBlock(summary, painter, glyphs) {
  const total = summary.succeeded + summary.skipped + summary.failures.length;
  const rule = glyphs.rule.repeat(HEADER_WIDTH);
  const counts = [
    total + ' ' + plural(total, 'repository', 'repositories') + ', ' + summary.succeeded + ' succeeded',
  ];
  if (summary.failures.length > 0) counts.push(summary.failures.length + ' failed');
  if (summary.skipped > 0) counts.push(summary.skipped + ' skipped');

  let out = painter.paint('cyan', rule) + '\n';
  const style = summary.failures.length === 0 ? 'boldGreen' : 'red';
  out += painter.paint(style, 'mgit: ' + counts.join(', ')) + '\n';
  for (const failure of summary.failures) {
    out += '  ' + painter.paint('red', glyphs.failure + ' ' + failure.name) +
      ' (exit code ' + failure.exitCode + ')\n';
  }
  return out;
}

/** The message that is printed when no repository was found. */
export function missingRepositoryWarning(root, depth) {
  return 'mgit: no Git repository found in `' + root + '` (searched ' + depth + ' ' +
    plural(depth, 'level', 'levels') + ')\n';
}

/** The message that is printed when the git executable cannot be started. */
export function missingGitMessage(program, platform) {
  const hint = platform === 'windows'
    ? '       Install it with `winget install --id Git.Git` or from https://git-scm.com/download/win.\n'
    : platform === 'macos'
      ? '       Install it with `xcode-select --install` or `brew install git`.\n'
      : '       Install it with your package manager, for example `sudo apt install git`.\n';
  return 'mgit: cannot run `' + program + '`: git was not found\n' + hint;
}

function plural(count, singular, many) {
  return count === 1 ? singular : many;
}

/** Remove every escape sequence, for copying the output as plain text. */
export function stripAnsi(text) {
  return text.replace(new RegExp(ESC + '\\[[0-9;]*m', 'g'), '');
}
